### ISSUE TITLE
Order of evaluation of multiple expressions in class CustomManyParticleForce

### ISSUE THREAD

**Issue by weiHelloWorld** [pandegroup/openmm#1141]
Order of evaluation of multiple expressions in class CustomManyParticleForce
==========================================

Hi,

I am working with OpenMM and would like to add a custom force to the system using class CustomManyParticleForce.  

And I found that when the custom potential energy is expressed in multiple expressions, there is a default order of evaluation: evaluating from the last one to the first one.  And whenever it sees a variable undefined in this evaluation process, it simply treats it as some "constant potential" (which means it would not exert any forces to the system).  

To be more clear, I would show you following example:

```python
# simulation (add CustomManyParticleForce)


from simtk.openmm.app import *
from simtk.openmm import *
from simtk.unit import *
from sys import stdout

############################ PARAMETERS BEGIN ###############################################################
force_constant = sys.argv[1] # read force constant from command line argument

record_interval = 50 
total_number_of_steps = 1000

input_pdb_file_of_molecule = 'alanine_dipeptide.pdb'
force_field_file = 'amber99sb.xml'
water_field_file = 'amber99_obc.xml'

pdb_reporter_file = 'biased_output_%s.pdb' %(force_constant)
state_data_reporter_file = 'biased_report_%s.txt' %(force_constant)

xi_1_0 = "-1"  # they  are the coordinates of center of the potential well
xi_2_0 = "-1"

k1 = force_constant
k2 = force_constant

energy_expression = '''
%s * (raw_layer_1_unit_2 - %s)^2 + %s * (raw_layer_1_unit_3 - %s)^2;  // line 0


dihedral_angle_2 = dihedral(p5, p7, p9, p15);  // line 1
raw_layer_1_unit_2 = cos(dihedral_angle_2);    // line 2
dihedral_angle_3 = dihedral(p7, p9, p15, p17); // line 3
raw_layer_1_unit_3 = cos(dihedral_angle_3);    // line 4

''' %(k1, xi_1_0, k2, xi_2_0)


############################ PARAMETERS END ###############################################################


pdb = PDBFile(input_pdb_file_of_molecule) 
forcefield = ForceField(force_field_file, water_field_file) # with water
system = forcefield.createSystem(pdb.topology,  nonbondedMethod=NoCutoff, \
                                 constraints=AllBonds)  # what does it mean by topology? 

# add custom force

force = CustomManyParticleForce(22, energy_expression) 
for i in range(system.getNumParticles()):
    force.addParticle("",0)  # what kinds of types should we specify here for each atom?
system.addForce(force)
# end add custom force
integrator = LangevinIntegrator(300*kelvin, 1/picosecond, 0.002*picoseconds)
integrator.setRandomNumberSeed(1)  # set random seed
simulation = Simulation(pdb.topology, system, integrator)
simulation.context.setPositions(pdb.positions)


simulation.minimizeEnergy()
simulation.reporters.append(PDBReporter(pdb_reporter_file, record_interval))
simulation.reporters.append(StateDataReporter(state_data_reporter_file, record_interval, step=True, potentialEnergy=True, kineticEnergy=True, totalEnergy=True, temperature=True))
simulation.step(total_number_of_steps)

print('Done!')
```

You can safely ignore all other parts and focus on the variable `energy_expression`.

If I write my expression this way, then the first few lines of simulation report would be

```
#"Step","Potential Energy (kJ/mole)","Kinetic Energy (kJ/mole)","Total Energy (kJ/mole)","Temperature (K)"
50,7873.91243594,6.18980425025,7880.10224019,35.4505453431
100,7880.86301732,10.1752390297,7891.03825635,58.2761195695
150,7884.42237776,15.831135263,7900.25351303,90.6688411758
200,7885.88277698,19.9950672577,7905.87784423,114.51671326
250,7892.28799962,26.3621868386,7918.65018646,150.982787514
300,7899.45199723,33.3880831654,7932.84008039,191.221839709
350,7906.17386277,38.8768699123,7945.05073268,222.657483808
400,7902.58416361,34.6754556932,7937.25961931,198.594941721
450,7903.76429835,43.9365031795,7947.70080153,251.635259406
500,7914.84469539,41.7024954429,7956.54719083,238.840542584
550,7899.69032373,51.2300255651,7950.92034929,293.40707247
600,7923.53266165,41.9639381483,7965.4965998,240.337889852
650,7910.49821964,53.7269511944,7964.22517083,307.707585323
700,7936.98418156,39.1858473942,7976.17002895,224.42707454
750,7924.6520872,54.1606645804,7978.81275178,310.191569538
800,7933.25161188,55.1419405482,7988.39355243,315.811580574
850,7949.20038345,49.453159595,7998.65354304,283.230520014
900,7924.00345396,65.3933964203,7989.39685038,374.524212918
950,7944.73851636,50.1203796072,7994.85889597,287.051854638
1000,7916.05481342,72.3908779324,7988.44569136,414.600526417
```

And the coordinates generated in this process are very close to those without this external potential, which means the expression simply adds a constant potential (about 8000 kJ/mol) to the system.

If I swap line 1 and line 2 in the `energy_expression`, then the report is

```
#"Step","Potential Energy (kJ/mole)","Kinetic Energy (kJ/mole)","Total Energy (kJ/mole)","Temperature (K)"
50,3878.79043905,5.87176936044,3884.66220841,33.6290805882
100,3886.54793804,9.49953221218,3896.04747026,54.4061789051
150,3891.11187168,15.1508281768,3906.26269986,86.7725536307
200,3894.34085326,22.0737831543,3916.41463642,126.422035167
250,3902.6297844,24.4433080328,3927.07309243,139.992892298
300,3914.60943548,31.373723593,3945.98315907,179.685102444
350,3907.87500436,40.9405810808,3948.81558544,234.476869914
400,3908.01087406,36.9571295489,3944.96800361,211.662654239
450,3914.23234127,43.1206010422,3957.35294231,246.962385347
500,3929.08515482,41.1014317513,3970.18658657,235.398101629
550,3915.72029716,55.8066892059,3971.52698637,319.618761136
600,3929.59416053,51.5833702169,3981.17753075,295.430764996
650,3913.78391764,61.9380766499,3975.72199429,354.73473893
700,3936.49364168,39.4805330651,3975.97417475,226.114812522
750,3909.65392324,74.119456149,3983.77337939,424.500522923
800,3944.19456585,41.3447697975,3985.53933565,236.79175902
850,3913.11885666,68.4454269517,3981.56428361,392.003949331
900,3931.38773171,47.8959798877,3979.2837116,274.312165315
950,3933.10576173,54.0648138982,3987.17057563,309.64260889
1000,3915.93000533,59.7520790843,3975.68208441,342.214988275
```

By analyzing the coordinates generated, I see that the first part of line 0 works for the system, while the second part still becomes a constant potential (about 4000 kJ/mol).

If I also swap line 3 and line 4, the report is

```
#"Step","Potential Energy (kJ/mole)","Kinetic Energy (kJ/mole)","Total Energy (kJ/mole)","Temperature (K)"
50,-119.905786702,5.56694921962,-114.338837482,31.8832999808
100,-112.46206961,10.0979570104,-102.3641126,57.8335062625
150,-105.754164825,13.9948928162,-91.7592720083,80.1522249001
200,-104.63523212,22.8226135319,-81.812618588,130.710772611
250,-97.828435517,24.6079505535,-73.2204849635,140.935840881
300,-89.7496899758,35.1461214417,-54.6035685341,201.29056129
350,-80.5209464383,30.0169602799,-50.5039861584,171.914582182
400,-93.1121897937,41.971573543,-51.1406162507,240.381619653
450,-81.5255916704,35.3802891853,-46.1453024851,202.631698081
500,-91.3903617409,44.9071770571,-46.4831846838,257.194549639
550,-78.1117084938,28.6620857308,-49.449622763,164.154879339
600,-77.3069006906,33.0373922966,-44.269508394,189.213346057
650,-94.2056982049,49.2229106056,-44.9827875994,281.9118269
700,-81.1231751849,37.0520044179,-44.071170767,212.206026163
750,-84.9816367982,47.6315906087,-37.3500461894,272.797942289
800,-79.6799146251,42.4497863264,-37.2301282987,243.120462963
850,-91.1574286957,41.938602822,-49.2188258737,240.19278815
900,-93.0458813259,48.216532051,-44.8293492749,276.14804712
950,-83.2197264882,51.4047963317,-31.8149301565,294.408028031
1000,-83.6297233198,49.8224547242,-33.8072685956,285.345564884
```

And everything is exactly what I expected.

In short, order of expressions really matters in this simulation, and if I make it wrong, I did not get errors or warnings.  I do not know whether others have similar issues.  I am wondering would it be better if you throw an exception in this case or parse the string of expression in another way (e.g. using multiple scans so that users do not need to care about order of expressions)?

Thanks in advance!

**Comment by peastman** [comment]
It's certainly supposed to throw an exception if anything in the expression is undefined.  Which platform are you using?

**Comment by weiHelloWorld** [comment]
I am using CPU platform on Linux.

**Comment by peastman** [comment]
It looks like the CUDA and OpenCL platforms validate all the variables, but the CPU and reference platforms don't.  I'm working on a fix right now.

**Comment by weiHelloWorld** [comment]
OK. Thanks for fixing it!

**Comment by peastman** [comment]
I merged the fix.
